### PR TITLE
sizeoferror for struct and sequence #771

### PR DESCRIPTION
**Pull request by jpsnyder**
- Fixes SizeOfError as described by #771 
- Created `Context` object for better organization
- Improved `Array` sizeof() function.

### CODE REVIEW

**Comment by arekbulski** [comment]
I wont accept a PR that I am not able to read entirely... but I am not closing it either. Perhaps it requires further study. Thank you for collaboration either way!

**Comment by jpsnyder** [comment]
Okay, thanks! Let me know if you have any questions.

**Comment by arekbulski** [comment]
Is there a reason why Travis didnt test this one?

**Comment by jpsnyder** [comment]
I'm not quite sure... It tested fine when I ran pytest locally though.

**Comment by jpsnyder** [comment]
Looks like Travis reports construct as an inactive repo: https://travis-ci.org/construct/construct
Perhaps you need to reactivate?

**Comment by arekbulski** [comment]
I logged into Travis (admin) pages, but there is no reactivate menu there...

**Comment by jpsnyder** [comment]
I'm not sure then... I don't know much about how Travis CI works.

**Comment by arekbulski** [comment]
Seems related if not the fix itself:
https://github.com/travis-ci/travis-ci/issues/5629

File: construct/core.py [hunk 1]
```diff
@@ -69,6 +69,61 @@ class CancelParsing(ConstructError):
...
+        context._root = self._root or context
+        return context
+
+    def get_child(self, name, default=None):
+        """Retrieves child Context or returns default if it doesn't exist."""
+        child_context = self.get(name, None)
+        if child_context and isinstance(child_context, Context):
+            return child_context
+        else:
+            return default
```
Comment (arekbulski) [inline construct/core.py]: This will produce None when the child value is falsey eg. False or [ ] for example.

File: construct/core.py [hunk 2]
```diff
@@ -2263,7 +2307,11 @@ def _sizeof(self, context, path):
                 count = count(context)
         except (KeyError, AttributeError):
             raise SizeofError("cannot calculate size, key not found in context")
-        return count * self.subcon._sizeof(context, path)
+        size = 0
+        for i in range(count):
+            context._index = i
+            size += self.subcon._sizeof(context, path)
+        return size
```
Comment (arekbulski) [inline construct/core.py]: This should be a separate Issue entirely.

File: tests/test_core.py [hunk 3]
```diff
@@ -421,6 +421,20 @@ def test_struct_issue_566():
...
+            'count' / Int32ul,
+            'entries' / Byte[this.count]
+        )
+    )
+    data = b'\x01\x00\x00\x00\x02\x00\x00\x00\x0a\x0b'
+    info = spec.parse(data)
+    assert info == {'a': 1, 'b': {'count': 2, 'entries': [0x0a, 0x0b]}}
+    assert spec.build(info) == data
+    assert spec.sizeof(**info) == 10
+
```
Comment (arekbulski) [inline tests/test_core.py]: Already added to testsuite.

File: construct/core.py (hunk 1 above)
Comment (jpsnyder) [reply construct/core.py:124]: Shouldn't that be expected? A child context should be of type `Context`, and be a truthy value even if the context hasn't been filled. Returning `None` indicates the context has no child.

(sorry for dropping the ball on this, btw)

File: construct/core.py (hunk 2 above)
Comment (jpsnyder) [reply construct/core.py]: Moved to new issue: 
https://github.com/construct/construct/issues/877

File: tests/test_core.py (hunk 3 above)
Comment (jpsnyder) [reply tests/test_core.py]: removed

**Comment by jpsnyder** [comment]
@arekbulski I have updated this pull request based on your code review. Please take a look again at your convenience.

**Pull request closed without merging**